### ISSUE TITLE
Enable deleting the last non-generated term in common rules rewriter

### ISSUE THREAD

**Issue by JohannesDaniel** [querqy/querqy#94]
Currently, the common rules rewriter will not delete a term if the query will not contain at least one non-generated term in the query after the deletion. 

This behavior leads to the following issues:
- The behavior is quite counterintuitive.
- It is not possible to redefine a query by replacing its terms by a combination of boost and filter queries. This would be very powerful as filter queries allow the definition of raw queries, which allows access to the big universe of Solr/Elasticsearch search capabilities.

By deleting the last non-generated term, the rewriter should show the following behavior:
- If the query still contains generated terms, the query will be processed in the existing way using these terms.
- If the query does not contain any term anymore, but contains boost and/or filter queries, the query will be defined as a MatchAllQuery.
- If the query neither contains terms nor boost or filter queries anymore, the query will be defined as NoMatchQuery returning 0 results.

**Comment by renekrie** [comment]
Makes sense to me. It might be tricky to cover the second case: 

> If the query does not contain any term anymore, but contains boost and/or filter queries, the query will be defined as a MatchAllQuery.

Let's assume we are left with a query `term1` and want to apply the following rule:

```
term1 =>
  DELETE: term1
  UP(100): term2
```

`term2` might not have been added when we apply the DELETE. I think the simplest solution would be to sort the instructions (DELETE, UP, SYNONYM etc.) per input and to put DELETE at the end.

**Comment by JohannesDaniel** [comment]
It is a bit more complicated. The current implementation considers not to remove dmqs, boolqs and the like. There might be additional rules basing on the same term that is supposed to be deleted, e. g. 

`a b => 
  DELETE: a

a => 
  SYNONYM: c
`

From a user perspective I would expect query `a b` to become `q(dmq(c), dmq(b))`.
This is why I consider to refactor DeleteInstruction in a way that it only collects terms to be deleted, but does not delete terms directly. Collected terms are deleted after all instructions are applied. The model is furthermore enhanced by a delete traversal, traversing the tree and removing all clauses that are empty after a term has been deleted.

**Comment by JohannesDaniel** [comment]
Thank you for your input. I see even more complications here. There might be additional rules basing on the same term that is supposed to be deleted, e. g. 

```
a b => 
  DELETE: a

a => 
  SYNONYM: c
	
a =>
	UP(1.0): b
```

From a user perspective I would expect query `a b` to become `q(dmq(c), dmq(b))`. Currently, the query simply becomes q(dmq(b)). This is why I consider to refactor DeleteInstruction in a way that it only collects terms to be deleted, but does not delete terms directly. Collected terms are deleted after all instructions have been applied. The model is furthermore enhanced by a delete traversal, traversing the tree and removing all clauses that are empty after a term has been deleted.

**Comment by JohannesDaniel** [comment]
Hmm, but the user could simply move the second rule to the top. So the current solution is even more flexible.

**Comment by renekrie** [comment]
Fixed by PR #95
